src/codegen.cpp

```diff
@@ -8133,17 +8135,7 @@ jl_llvm_functions_t jl_emit_code(
         "functions compiled with custom codegen params must not be cached");
     JL_TRY {
         decls = emit_function(m, li, src, jlrettype, params);
-        auto stream = *jl_ExecutionEngine->get_dump_emitted_mi_name_stream();
```

Comment:
This "dereference" is actually overloaded to take the stream's lock (defined in `jl_locked_stream` in jitlayers.h) and then release when the scope block ends.

Reply:
fascinating. that was super non-obvious from the call site, but that's really good to know!

Reply:
Thanks Prem!

Reply:
Note that this dereference locking is also used in the debuginfo code as well when interacting with the global registry.